k2/csrc/torch_api.h

```diff
@@ -21,6 +21,13 @@
 
 #include <memory>
 
+// TODO(WeiKang): Move the following lines to k2/csrc/all.h
+#include "k2/csrc/fsa_algo.h"
+#include "k2/csrc/fsa_utils.h"
+#include "k2/csrc/ragged.h"
+#include "k2/csrc/ragged_ops.h"
+#include "k2/csrc/rnnt_decode.h"
+#include "k2/csrc/torch_util.h"
 #include "torch/script.h"
```

Comment:
@csukuangfj  Why we need this extra torch wrapper, I think we can do it just like pytorch by putting the header files in one header file (like all.h), no need to wrap them again.

Reply:
Do you need NVCC to compile sherpa if you do it this way?

The original design tries to achieve the goal that whether you use a CPU version of k2 or a CUDA version as a pre-compiled library, you can always use GCC to compile sherpa.

Reply:
I try compile sherpa with CPU version Pytorch and CUDA version k2, it compiles successfully and runs normally.

Reply:
> I tried compiling sherpa with CPU version Pytorch and CUDA version k2, it compiles successfully and runs normally.

Do you have CI tests to cover that?

It looks surprising to me that k2 headers can be compiled with GCC when CUDA is enabled as there are stuffs that can only be processed by NVCC, e.g., kernels.

Reply:
Can you show the output of

make VERBOSE=1


You have to run 

make clean

Before you run "make VERBOSE=1".

Reply:
> I tried compiling sherpa with CPU version Pytorch and CUDA version k2, it compiles successfully and runs normally.

Have you tried to compile sherpa and k2 with CUDA support?

---

There are header files in k2 that use different macros depending on whether NVCC is present.
For instance,

https://github.com/k2-fsa/k2/blob/6a899805cacf82e69953e591e8683951d167fa26/k2/csrc/context.h#L44-L46


In your case, even if it does not throw compile-time errors, it does not mean it can run correctly.

The headers of k2 that you include in sherpa are compiled with different flags from those used by pre-compiled k2 libraries.

In short, the header files of k2 that you are using in sherpa are always for CPU, not for CUDA.
That is, the same template may be instantiated differently in sherpa and k2 if k2 is using CUDA, which I think
can cause undefined behavior during link time.

Reply:
> There are header files in k2 that use different macros depending on whether NVCC is present.

I know what you said.

> In your case, even if it does not throw compile-time errors, it does not mean it can run correctly.

At least, it runs normally when decoding with fast_beam_search.

Will do more experiments.

Reply:
If I use cuda version Pytorch, how could I make sherpa compile with gcc?

Reply:
It may depend what functions we use.  If we use certain functions from a .cc file compiled with gcc that would normally directly instantiate a CUDA kernel, it may compile something that will just generate an error instead of really using CUDA.